File: src/main/java/com/ryanpmartz/kupload/service/UploadService.java
package com.ryanpmartz.kupload.service;

import com.amazonaws.services.s3.model.CannedAccessControlList;
import com.amazonaws.services.s3.model.ObjectMetadata;
import com.amazonaws.services.s3.model.PutObjectRequest;
import com.amazonaws.services.s3.transfer.TransferManager;
import org.apache.log4j.Logger;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.stereotype.Service;
import org.springframework.web.multipart.MultipartFile;

import java.io.IOException;
import java.io.InputStream;
import java.util.UUID;

@Service
public class UploadService {

    private static final Logger log = Logger.getLogger(UploadService.class);

    private TransferManager transferManager;
    private String s3Bucket;

    @Autowired
    public UploadService(TransferManager transferManager, @Value("s3.bucket.name") String s3Bucket) {
        this.transferManager = transferManager;
        this.s3Bucket = s3Bucket;
    }

    public void uploadFile(MultipartFile multipartFile) {
        String extension = determineExtension(multipartFile);
        String filename = UUID.randomUUID().toString() + extension;

        try (InputStream is = multipartFile.getInputStream()) {

            PutObjectRequest req = new PutObjectRequest(s3Bucket, filename, is, new ObjectMetadata());
            req.withCannedAcl(CannedAccessControlList.Private);

            transferManager.upload(req);

        } catch (IOException e) {
            log.error("IOException occurred uploading file", e);
            throw new RuntimeException(e);
        }
    }

    private String determineExtension(MultipartFile multipartFile) {
        String filename = multipartFile.getOriginalFilename();

        return filename.substring(filename.lastIndexOf(".") + 1);
    }
}

File: src/main/java/com/ryanpmartz/kupload/service/UploadService.kt
package com.ryanpmartz.kupload.service

import com.amazonaws.services.s3.AmazonS3Client
import com.amazonaws.services.s3.model.CannedAccessControlList
import com.amazonaws.services.s3.model.ObjectMetadata
import com.amazonaws.services.s3.model.PutObjectRequest
import org.apache.log4j.Logger
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.beans.factory.annotation.Value
import org.springframework.stereotype.Service
import org.springframework.web.multipart.MultipartFile
import java.util.*

@Service
class UploadService @Autowired
constructor(private val s3Client: AmazonS3Client, @Value("\${aws.s3.bucket}") private val s3Bucket: String) {

    fun uploadFile(multipartFile: MultipartFile): String {
        val extension = determineExtension(multipartFile)
        val filename = UUID.randomUUID().toString() + "." + extension

        multipartFile.inputStream.use { mis ->
            val req = PutObjectRequest(s3Bucket, filename, mis, ObjectMetadata())
            req.withCannedAcl(CannedAccessControlList.Private)

            s3Client.putObject(req)
        }

        return filename

    }

    private fun determineExtension(multipartFile: MultipartFile): String {
        val filename = multipartFile.originalFilename

        return filename.substring(filename.lastIndexOf(".") + 1)
    }

    companion object {

        private val log = Logger.getLogger(UploadService::class.java)
    }
}
